doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,279 @@
+cluster. Especially if the initial scope does not include `SELECT` queries which
+would have required spinning up compute dataflows.
+Reasons to not do so,
+- As mentioned above, we will eventually support `SELECT` queries in `COPY ... TO S3`. The
+only reason it would have made sense to do this in storage would be to do as a quick stop gap
+if the effort is drastically less. It does not appear to be so.
+- Unlike Compute, Storage does not have a concept of running temporary/one-off dataflows.
+Everything which runs, like sources or sinks, do so continuously once scheduled till they are
+dropped.
+
+### Output file format
+We should start with a single file format and add more if there's more user interest. Currently,
+there has been request for parquet but CSV support would be easiest to build.
+
+Postgres supports `COPY ... TO` with a csv file and we should mimic the file structure for our
+export as well.
+
+Eventually we'll need to add some additional formatting options to the SQL like
+`DELIMITER` (default will be comma), `QUOTE` (default is double quotes) etc. In the initial
+version we should stick to the default values as per the
+[Postgres COPY command](https://www.postgresql.org/docs/current/sql-copy.html).
+
+### Bounded Memory
+I had wrongly assumed that with reading directly from persist we could use backpressure similar to what we do in upsert. As @guswynn correctly pointed out, that wouldn't work due to the following:
+* For upsert, we read configured amount of data from persist, but we eventually consolidate in rocksdb (or in memory if disk is not enabled) which maintains the upsert state.
+* In compute, we may be able to fetch data in chunks from persist, but it will still need to be
+consolidated for the entire shard and this will have to happen in memory.
+
+One possibility could be that we actually use disk (with rocksdb) in Compute to keep state similar
+to upsert and consolidate the data there before writing the snapshot to S3. This seems complicated
+and I don't think it will be worth the effort. Also, given that we might
+have arrangements spilling to disk in the future, we'll probably get this for free.
+
+TODO: Check on the status of consolidation on read from persist, if it can be used here.
+
+Since backpressure is not feasible right off the bat when reading from materialized view as I had
+assumed, there's no reason to limit the scope to only support materialized views/tables and
+we can support arbitrary SELECT queries as well.
+
+### Writing to S3
+While writing to S3, we'll need to split the data into multiple files so that each file is
+under the configured MAX FILE SIZE. In the initial version, we can have a single timely worker
+doing the [multi part uploads](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)
+to S3.
+* A single worker starts a multi part upload with a `CreateMultipartUpload` API call to start
```

Comment:
I'm assuming there is a hidden parameter here which defines the maximum amount of data that we're willing to store in memory before creating a part which might be less than the max file size, right? AWS suggests that this should be around the 100MB mark so maybe we can spell it out in the design doc.

Reply:
Makes sense, will add.

Reply:
Added.